### PR TITLE
Align resource loading logic in `AntClassLoader` with upstream

### PR DESCRIPTION
**Pull request by basil**
Our customizations to `AntClassLoader` span two classes: `jenkins.util.AntClassLoader` and its subclass `jenkins.util.AntWithFindResourceClassLoader`. The reason for this split is historical: the logic in `jenkins.util.AntWithFindResourceClassLoader` began as a subclass of `org.apache.tools.ant.AntClassLoader` before `jenkins.util.AntClassLoader` came to exist. In this context, subclassing made sense as a way to avoid code duplication. But at some point `org.apache.tools.ant.AntClassLoader` was copied into the source tree as `jenkins.util.AntClassLoader`. At that point subclassing stopped making sense, but this was never cleaned up. And in 2021, this makes it harder for us to upstream changes to our fork and to merge those upstream changes back to our fork. So this PR does two things:

- Merges `jenkins.util.AntClassLoader` and its subclass `jenkins.util.AntWithFindResourceClassLoader` into one class, which is a copy of the upstream `org.apache.tools.ant.AntClassLoader` class with some Jenkins-specific modifications.
- Where those Jenkins-specific modifications have been upstreamed recently, prefer the upstream fix to our own fix.

The tricky part of this change is the last part, so it's time for a history lesson.

### Background

In 2004, Java 1.4 reigned supreme and Kubernetes was yet to roam the face of the earth. Now recall that `ClassLoader#getResources(String name)` does a recursive search for resources, starting by delegating to its parent, then by delegating to the bootstrap class loader, then by searching its own path(s) via `ClassLoader#findResources(String name)`. So `ClassLoader#findResources(String name)` is effectively the base case of the recursion for `ClassLoader#getResources(String name)`. `ClassLoader#getResources(String)` was `final` in this version of Java, which proved to be a serious issue for subclasses such as `org.apache.tools.ant.AntClassLoader`. Yes, they could override `ClassLoader#findResources(String name)` in the subclass, but the only way for the subclass's version to be invoked was via the parent class's `ClassLoader#getResources(String name)`, which searched its own parent first before delegating to the subclass's `ClassLoader#findResources(String name)`.

Complicating this issue for `AntClassLoader` was that it maintained its own `parent` field independently of its parent class `ClassLoader`'s `parent` field. Unlike its parent class `ClassLoader`, whose `parent` field is `final`, `AntClassLoader` maintains a mutable `parent` field. `AntClassLoader#findResources(String name)` did not behave properly when `AntClassLoader#parent` was something other than `ClassLoader#parent`. This led an intrepid NetBeans developer named Jesse Glick to file [Ant bug 30161](https://bz.apache.org/bugzilla/show_bug.cgi?id=30161) on July 16, 2004, along with a patch to correct `AntClassLoader`'s behavior in the common case. The patch came with the following comment:

> Note that there is a problem with the patch (which is I believe unavoidable unless you set `CL.parent` correctly, or override `CL.getResources` in JDK 1.5 where it seems to be nonfinal): if `CL.parent` (whatever that is) does have any matches for the resource name, and `ACL.parent` has matches for it that are in fact derived ultimately from `CL.parent`, `ACL.getResources` will produce duplicates. I put in a quick hack to avoid this in case `CL.parent == ACL.parent`, but for a more general fix I suppose you would have to call `CL.parent.getResources` and remove all such matches before returning them from `ACL.findResources`. I have not tried such a refinement yet.

In other words, the patch contained a known issue that could only be solved by `AntClassLoader` setting `ClassLoader#parent` or overriding `ClassLoader#getResources(String)` (which wasn't yet possible in Java 1.4). The patch contained a workaround for the common case where `ClassLoader.parent == AntClassLoader.parent`, which seemed to suffice for most use cases in the short term.

* * *

Meanwhile, confusion remained about the two separate `parent` fields (the immutable one in `ClassLoader` and the mutable one in `AntClassLoader`). This was discussed in [Ant bug 35436](https://bz.apache.org/bugzilla/show_bug.cgi?id=35436), where several options were proposed. The most desirable option from a code cleanliness perspective was to make `AntClassLoader`'s `parent` field immutable and to set the `parent` field on `ClassLoader` to the same value. However, this option was less desirable from a backward compatibility perspective, as it would have been a breaking change for many consumers. Ultimately, this proposal was rejected on July 22, 2009 by Stefan Bodewig, who wrote:

> I don't think we would achieve much by setting `ClassLoader.parent` since we'd still have to deal with `parentFirst` and `setIsolated` in a way incompatible with the base implementation. In order to close this issue I've simply added a `getConfiguredParent` method in svn revision 796649

In other words, Ant made a clear decision to maintain a separate `parent` field from its parent class, even going so far as to add a second getter. The original `getParent()` getter still returned the `parent` field from `ClassLoader`, while the new `getConfiguredParent()` getter returned the `parent` field from `AntClassLoader`. This obviously isn't ideal from a code cleanliness perspective, but it does work from a backward compatibility perspective.

* * *

Let's get back to Jesse Glick's fix for [Ant bug 30161](https://bz.apache.org/bugzilla/show_bug.cgi?id=30161) on July 16, 2004. Recall that it contained a known issue that was worked around in the common case. Fast forward 10 years and this code, now being used in Jenkins rather than NetBeans, started causing some problems on March 25, 2014 in [JENKINS-22310](https://issues.jenkins.io/browse/JENKINS-22310?focusedCommentId=197405&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-197405). The workaround for the common case where `ClassLoader#parent == AntClassLoader#parent` did not apply to Jenkins, and the abovementioned issue was visible in the form of half-second class lookups.

Never one to shy away from sinking his teeth into a tough technical problem, Kohsuke Kawaguchi debugged the issue and found the problematic line in `AntClassLoader`. In retrospect, we know that there were two possible solutions: setting `ClassLoader#parent` (which would imply changing the semantics of `AntClassLoader` to make the parent immutable) or overriding `ClassLoader#getResources` in JDK 1.5 where is non-`final`. Kohsuke chose to fix this the way that seemed natural to him, which was to set `ClassLoader#parent` to `AntClassLoader#parent`. Kohsuke added this logic:

```java
@@ -284,7 +284,6 @@ public class AntClassLoader extends ClassLoader implements SubBuildListener, Clo
      * @param classpath The classpath to use to load classes.
      */
     public AntClassLoader(final ClassLoader parent, final Project project, final Path classpath) {
-        super(parent);  // KK patch for JENKINS-21579
         setParent(parent);
         setClassPath(classpath);
         setProject(project);
@@ -371,7 +370,6 @@ public class AntClassLoader extends ClassLoader implements SubBuildListener, Clo
      *                    load the a class through this loader.
      */
     public AntClassLoader(final ClassLoader parent, final boolean parentFirst) {
-        super(parent);  // KK patch for JENKINS-21579
         setParent(parent);
         project = null;
         this.parentFirst = parentFirst;
```

(Here "KK" stands for "Kohsuke Kawaguchi.")

This solved the problem introduced by the original workaround in Jesse Glick's 2004 patch, but it went against the direction Stefan Bodewig took in the 2009 patch that added `getConfiguredParent()`. In retrospect, it is not entirely clear that Kohsuke Kawaguchi was aware of Stefan Bodewig's change or the original context for Jesse Glick's 2004 workaround. So Kohsuke's change is perfectly understandable. Nevertheless, it is not ideal, and this has code survived until the present day.

* * *

When I went to upstream Kohsuke's patch, I received a less than enthusiastic response. And rightly so: in the context of [Ant bug 35436](https://bz.apache.org/bugzilla/show_bug.cgi?id=35436), Kohsuke went in the opposite direction of the one already taken by the Ant developers. It would have been better to undo Jesse Glick's original workaround and implement the solution originally described by Jesse Glick in 2004: overriding `ClassLoader#getResources`, which had by then been non-`final` for a decade.

It turns out that during the intervening years Ant had already overridden `ClassLoader#getResources` as part of a miscellaneous code cleanup change, so the only thing left to do was remove Jesse Glick's 2004 workaround and Kohsuke's 2014 patch. I did the former in apache/ant#151 and am doing the latter in this PR.

With those two tasks completed, this 17-year-old piece of technical debt has been paid off.

### Testing done

- Used `usage-in-plugins` to confirm there were no references to `jenkins.util.AntWithFindResourceClassLoader` anywhere in the ecosystem.
- Reverted the "KK patch for JENKINS-21579" on the main branch and reproduced the original problem (half-second class lookups) with Kohsuke's script console code from [JENKINS-22310 (comment)](https://issues.jenkins.io/browse/JENKINS-22310?focusedCommentId=197405&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-197405). Then ran the same test with the changes in this PR (including the upstream fix from apache/ant#151 and Kohsuke's patch reverted) and confirmed that the issue could no longer be reproduced.

### Future work

Once I backport further upstream changes to our fork, and once my upstream changes to `AntClassLoader` are released and Jenkins upgrades to them, it might be possible to make `jenkins.util.AntClassLoader` again a subclass of `org.apache.tools.ant.AntClassLoader` and eliminate most of the duplicate code. That would help decrease our maintenance burden, but we aren't ready for that yet. This PR gets us closer to that goal, though.

### Proposed changelog entries

Developer: Jenkins now uses an updated version of the `AntClassLoader` class with fewer custom patches.

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [x] (If applicable) Jira issue is well described
- [x] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [x] Appropriate autotests or explanation to why this change has no tests
- [x] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@mention

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are correct
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins-ci.org/issues/?filter=12146)).

### CODE REVIEW

File: core/src/main/java/jenkins/util/AntClassLoader.java
Comment (basil) [inline core/src/main/java/jenkins/util/AntClassLoader.java:287]: No longer needed as of apache/ant#151.

File: core/src/main/java/jenkins/util/AntClassLoader.java
Comment (basil) [inline core/src/main/java/jenkins/util/AntClassLoader.java:374]: No longer needed as of apache/ant#151.

File: core/src/main/java/jenkins/util/AntClassLoader.java
Comment (basil) [inline core/src/main/java/jenkins/util/AntClassLoader.java:520]: A Jenkins-specific change, not present upstream. Moved from `AntWithFindResourceClassLoader` without modification.

File: core/src/main/java/jenkins/util/AntClassLoader.java
Comment (basil) [inline core/src/main/java/jenkins/util/AntClassLoader.java:909]: I cleaned up and upstreamed this change in apache/ant#150. This just adjusts our local copy to match the final version that was accepted upstream.

File: core/src/main/java/jenkins/util/AntClassLoader.java
Comment (basil) [inline core/src/main/java/jenkins/util/AntClassLoader.java:949]: I cleaned up and upstreamed this change in apache/ant#150. This just adjusts our local copy to match the final version that was accepted upstream.

File: core/src/main/java/jenkins/util/AntClassLoader.java
Comment (basil) [inline core/src/main/java/jenkins/util/AntClassLoader.java:978]: Bringing in the upstream change from apache/ant#150.

File: core/src/main/java/jenkins/util/AntClassLoader.java
Comment (basil) [inline core/src/main/java/jenkins/util/AntClassLoader.java:1010]: Bringing in the upstream change from apache/ant#151.

File: core/src/main/java/jenkins/util/AntClassLoader.java
Comment (basil) [inline core/src/main/java/jenkins/util/AntClassLoader.java:1611]: Another upstream change being pulled in.

File: core/src/main/java/jenkins/util/AntClassLoader.java
Comment (basil) [inline core/src/main/java/jenkins/util/AntClassLoader.java:1605]: A Jenkins-specific change moved without modification from `AntWithFindResourceClassLoader`.

**Review by timja — APPROVED** [review]

**Review by res0nance — APPROVED** [review]

**Comment by timja** [comment]
This PR is now ready for merge, after ~24 hours, we will merge it if there's no negative feedback.

Thanks!

**Review by daniel-beck — APPROVED** [review]
Looks good. Thanks for the history lesson, very well researched!

I reviewed the diff between upstream and this and it seems there's mostly trivial changes and some visibility increases left. FWIW `MissingClassTelemetry` is no longer used, so that may help a bit as well. So AFAICT we're _very_ close to that goal, but I may be missing something.

**Comment by basil** [comment]
> I reviewed the diff between upstream and this and it seems there's mostly trivial changes and some visibility increases left.

Indeed. The visibility changes are actually required for some performance optimizations Kohsuke added to Jenkins plugin loading. Kohsuke implemented them with reflection, but with the advent of Java 11 that prints a reflective access warning so Tim had to increase the visibility level. I think this approach could still be maintained even if we end up making `jenkins.util.AntClassLoader` a subclass of `org.apache.tools.ant.AntClassLoader`.

There's still one more upstream change to call `registerAsParallelCapable()`. I'm waiting for this to be merged and released before I open a PR with that change. In general I've been trying to stagger these Ant changes over several weeks so that in case anything goes wrong bisection will be easier. Calling `registerAsParallelCapable()` might even lead to a small performance improvement.

> FWIW `MissingClassTelemetry` is no longer used, so that may help a bit as well.

I wasn't quite sure what to do about `MissingClassTelemetry`. If it's really no longer used, I can remove the whole class and any references to it (essentially reverting #4003). That would get it out of the way here as well.

**Comment by daniel-beck** [comment]
> If it's really no longer used, I can remove the whole class and any references to it (essentially reverting #4003). That would get it out of the way here as well.

All telemetry is running for a defined, usually fairly short time and inactive afterwards. This one stopped earlier this year per https://github.com/jenkinsci/jenkins/blob/c7c3ae18ef7fe2b65f0c3ce4343cf26c66cfe475/core/src/main/java/jenkins/telemetry/impl/java11/MissingClassTelemetry.java#L69-L72

Unless we decide to restore it because we still need new data, this can probably be removed. Would check with folks involved in that effort to be safe.

**Comment by jglick** [comment]
> Calling `registerAsParallelCapable()` might even lead to a small performance improvement.

Possibly; see https://issues.jenkins.io/browse/JENKINS-23784

**Pull request merged**
